### PR TITLE
 plan: fix a bug when using correlated column as index

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

The for loop was wrongly breaked. Hence there may be case that we misjudged.

In the for loop, we check that whether this index column have corresponding equal condition which has correlated column.
So we should break once we found it. If one condition is not matched, we should continue to next condition rather than breaking the loop.

### What is changed and how it works?

Break the for loop at the correct position.

### Check List 

Tests 

 - Unit test

Will be added later.

Related changes

 - Need to be included in the release note

Fix the case that index may not use correlated column correctly.

### CODE REVIEW

File: plan/logical_plans.go [hunk 1]
```diff
@@ -484,14 +484,15 @@ func (path *accessPath) splitCorColAccessCondFromFilters() (access, remained []e
 	for i := path.eqCondCount; i < len(path.idxCols); i++ {
 		matched := false
 		for j, filter := range path.tableFilters {
-			if !isColEqCorColOrConstant(filter, path.idxCols[i]) {
-				break
+			if used[j] || !isColEqCorColOrConstant(filter, path.idxCols[i]) {
```
Comment (zz-jason) [inline plan/logical_plans.go]: It's better to add a UT for function `splitCorColAccessCondFromFilters` though we already have integration tests, with UT we can test some corner cases that are not easy to be covered.

Comment (zz-jason) [reply plan/logical_plans.go]: If we have UT for this function, even without the integration test you added in this PR, we can avoid this bug in the very beginning.

You can see the advantage of unit test through Wikipedia: https://en.wikipedia.org/wiki/Unit_testing#Advantages

File: plan/logical_plans_test.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+		return x, nil
+	}
+	return expr, nil
+}
+
+func (s *testUnitTestSuit) TestIndexPathSplitCorColCond(c *C) {
+	defer testleak.AfterTest(c)()
+	totalSchema := expression.NewSchema()
+	totalSchema.Append(&expression.Column{
+		ColName:  model.NewCIStr("a"),
```
Comment (zz-jason) [inline plan/logical_plans_test.go]: I think a better way is to:
```sh
s/a/col1/
s/b/col2/
s/c/col3/
s/d/col4/
s/e/col5/
```

It helps us understanding the test case.

File: expression/column.go [hunk 3]
```diff
@@ -375,3 +375,22 @@ func IndexInfo2Cols(cols []*Column, index *model.IndexInfo) ([]*Column, []int) {
...
+		found := false
+		for _, col := range cols {
+			if col.UniqueID == id {
+				retCols = append(retCols, col)
+				found = true
+				break
+			}
+		}
+		if !found {
+			break
```
Comment (zz-jason) [inline expression/column.go]: This check can be removed. we can make this function to find all the columns in the unique id list.

Comment (winoros) [reply expression/column.go]: But we still need this check. In such case, we need to return nil or a full slice?

Comment (zz-jason) [reply expression/column.go]: The function caller can check whether there is any column can not be found in `cols` by:

```go
len(ids) == len(retCols)
```

Comment (zz-jason) [reply expression/column.go]: Another consideration is, this function is only used in the test currently, all the column identified by `ids` can be found in `cols`. We can extend this function in the future when we meet more complicated scenarios. For now, I think we'd better to keep it simple.

**Comment by zz-jason** [comment]
@XuHuaiyu @lamxTyler PTAL

File: plan/logical_plans_test.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/pingcap/tidb/util/testleak"
+)
+
+var _ = Suite(&testUnitTestSuit{})
+
+type testUnitTestSuit struct {
+	ctx sessionctx.Context
+}
+
+func (s *testUnitTestSuit) SetUpSuite(col3 *C) {
```
Comment (lamxTyler) [inline plan/logical_plans_test.go]: Why name it `col3`?

File: plan/logical_plans_test.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		if _, ok := colIDs[x.UniqueID]; ok {
+			return &expression.CorrelatedColumn{Column: *x}, nil
+		}
+	default:
+		return x, nil
+	}
+	return expr, nil
+}
+
+func (s *testUnitTestSuit) TestIndexPathSplitCorColCond(col3 *C) {
```
Comment (lamxTyler) [inline plan/logical_plans_test.go]: ditto.

File: plan/logical_plans_test.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+			}
+			newArgs = append(newArgs, newArg)
+		}
+		newSf := expression.NewFunctionInternal(x.GetCtx(), x.FuncName.L, x.GetType(), newArgs...)
+		return newSf, nil
+	case *expression.Column:
+		if _, ok := colIDs[x.UniqueID]; ok {
+			return &expression.CorrelatedColumn{Column: *x}, nil
+		}
+	default:
```
Comment (crazycs520) [inline plan/logical_plans_test.go]: redundant default？ line#66 is enough.

**Comment by zz-jason** [comment]
@crazycs520 @XuHuaiyu PTAL

**Pull request merged**